### ISSUE TITLE
ocamlopt assertion failure when compiling prooftree

### ISSUE THREAD

**Issue by jamesjer** [ocaml/ocaml#9859]
Attempting to build [prooftree](http://askra.de/software/prooftree/) 0.13 with OCaml 4.11.0 ends with an assertion failure:
```
ocamlopt.opt -c -safe-string -w Aem-48 -g -I +lablgtk2 tree_layers.ml
Fatal error: exception File "typing/ctype.ml", line 4003, characters 6-12: Assertion failed
```
This did not happen with OCaml 4.10.0 and earlier.  By commenting out code and retrying the ocamlopt invocation, I was able to identify this as the code that triggers the assertion (lines 292-296 of tree_layers.ml):
```
  method add_layer l =
    layers <- layers @ [l];
    l#register_layer_stack (self :> tree_layer abstract_tree_container);
    self#clear_size_cache;
    List.length layers - 1 
```

**Comment by Octachron** [comment]
A quite simpler example (for our testsuite):
```ocaml
class setup = object
  method with_ f = (f 0:unit)
end
class fail = object (self)
  method trigger = (self :> setup )
end
```

**Comment by gasche** [comment]
@Octachron one downside of this example is that it is not a valid OCaml program as the `setup` coercion is incorrect. It would be nice to either duplicate the `with_` method in `fail` or add `inherit setup`; the resulting program is accepted with 4.10 and still raises an assertion failure in 4.11.

@jamesjer: thanks for the report! Note that we do a large part of our pre-release testing by checking that the program distributed in the Opam repository still compile. If you want the prooftree codebase to receive such pre-release testing, you could upload it as an opam package. (It would apparently have helped us for 4.11,)

**Comment by Octachron** [comment]
@gasche : the coercion is also correct for a virtual `fail` class.

**Comment by Octachron** [comment]
The assertion was introduced in #9348 in order to tentatively fix a strange information flow when an arrow type becomes spontaneously known.

**Comment by trefis** [comment]
FTR: I "looked" at this with @Octachron earlier today, my intuition is that the assertion is still warranted and I'm surprised that the arrow type we get at this point is "unknown":
```ocaml
{id=962;level=4;desc=
 Tarrow("",{id=964;level=4;desc=Tconstr(int,[],[])},
  {id=963;level=4;desc=Tconstr(unit,[],[])},Cunknown)}
```

Although it is consistent with the type we get for:
```
# let with_ f = (f 0 : unit);;
{id=1289;level=100000000;desc=
 Tarrow("",
  {id=1301;level=100000000;desc=
   Tarrow("",{id=1302;level=100000000;desc=Tconstr(int,[],[])},
    {id=1297;level=100000000;desc=Tconstr(unit,[],[])},Cunknown)},
  {id=1296;level=100000000;desc=Tconstr(unit,[],[])},Cok)}
```
(this is the `val_type` being printed)

Is this expected?

**Comment by jamesjer** [comment]
> @jamesjer: thanks for the report! Note that we do a large part of our pre-release testing by checking that the program distributed in the Opam repository still compile. If you want the prooftree codebase to receive such pre-release testing, you could upload it as an opam package. (It would apparently have helped us for 4.11,)

I am not directly connected with the prooftree project.  I maintain a number of formal methods packages for the Fedora Linux distribution (notably coq and friends), and prooftree is one of the packages I maintain.  I will try to contact prooftree upstream to see if there is interest in submitting to opam.

**Comment by lpw25** [comment]
> Is this expected?

I haven't looked at this aspect of the type checker in detail, but I think this behaviour is probably intended. It does give some weird behaviours, like:
```ocaml
# let foo f x y = f ~x ~y;;
val foo : (x:'a -> y:'b -> 'c) -> 'a -> 'b -> 'c = <fun>

# let bar g x y =
    let _ = foo g x y in
    g ~y ~x;;
Line 3, characters 4-5:
3 |     g ~y ~x;;
        ^
Error: This function is applied to arguments
       in an order different from other calls.
       This is only allowed when the real type is known.
```
vs.
```ocaml
# let foo (f : x:_ -> y:_ -> _) x y = f ~x ~y;;
val foo : (x:'a -> y:'b -> 'c) -> 'a -> 'b -> 'c = <fun>

# let bar g x y =
    let _ = foo g x y in
    g ~y ~x;;
val bar : (x:'a -> y:'b -> 'c) -> 'a -> 'b -> 'c = <fun>
```
but nevertheless I believe that is how they are expected to work.

**Comment by gasche** [comment]
But in your examples, `f` is applied with labels. In the original example, the application of `f` that gives an arrow type is `f 0`. Naively I would expect this to result `f` getting a *non-labelled* arrow type, and I see no reason to use Cunknown instead of Cok for non-labelled arrows.

**Comment by lpw25** [comment]
That's true, and I agree it is unnecessary to have Cunknown on an unlabelled parameter. However the assertion that is being triggered is for all function types, not just unlabelled ones, so it can still be triggered with a label:
```
# class setup = object
  method with_ f = (f ~x:0 : unit)
end
class fail = object (self)
  method trigger = (self :> setup )
end;;
Fatal error: exception File "typing/ctype.ml", line 4003, characters 6-12: Assertion failed
```

**Comment by garrigue** [comment]
> But in your examples, `f` is applied with labels. In the original example, the application of `f` that gives an arrow type is `f 0`. Naively I would expect this to result `f` getting a _non-labelled_ arrow type, and I see no reason to use Cunknown instead of Cok for non-labelled arrows.

No label is just the empty-string label, except when the function type is known and all arguments are provided. So, if the function type is unknown, no label should behave exactly as applying to the label empty-string.

**Comment by garrigue** [comment]
I see to solutions: either remove the assertion (which is not that useful), or change Cunknown to Cok when generating the class type (this kind of makes sense, as it is strange to leave artifacts of inference inside type definitions, but this is extra work).

**Comment by trefis** [comment]
>  either remove the assertion (which is not that useful)

I have no objections to removing the assertion, but I think saying it's not useful is an overstatement.
It was added because it seemed weird to arbitrarily change `Cunknown` to `Cok` at that point, and the only explanation one could come up with was that this wasn't actually happening: clearly a few of us were wrong on that case.

> change Cunknown to Cok when generating the class type (this kind of makes sense, as it is strange to leave artifacts of inference inside type definitions, but this is extra work).

I assume that if it makes sense for the class type, it also makes sense for the value type in @lpw25's example. As I understand it, you're refuting this statement:

> but I think this behaviour is probably intended

which I would find reassuring.

If that's right, I would be in favor of changing the `Cunknown` to `Cok` in all these cases (when generalising the type).

If the other "fix" is chosen (that is: removing the assertion), I would ask for a comment explaining why the code is OK.

**Comment by Octachron** [comment]
Fiddling with the issue, the problem appears as soon as one captures an inferred function type, which does not require going through the object system:
```ocaml
module type T = sig type t  end
let type_of (type x) (x: x) = (module struct type t = x end: T with type t = x)
let f g = 1 + g ~x:0 ~y:0;;
module E = (val type_of f)
let g = ( (fun _ -> f) :> 'a -> E.t)
```
> Fatal error: exception File "typing/ctype.ml", line 4062, characters 6-12: Assertion failed

**Comment by Octachron** [comment]
Since this is an isolated assertion which is contradicted by current behavior of the type checker, I am reverting #9348 in 4.11.1 . 

We can discuss the right behavior without breaking existing code.

@jamesjer : the specific prooftree issue should be fixed in 4.11.1, which will be released shortly to fix another major issue.

**Comment by jamesjer** [comment]
Thank you for the swift action on this issue.

**Comment by garrigue** [comment]
@trefis The meaning I was giving to "not useful" is that it is not needed to make the corresponding code work.
On the other hand, one could say that it was useful, since it revealed that the assertion was wrong.
But then making that an assertion (that causes fatal failures) might not have been a good idea.
If the goal is to have people report it, we could have a warning for that, but it's not clear it would work.

> I assume that if it makes sense for the class type, it also makes sense for the value type in @lpw25's example.

No, my idea was that a class type is also an object type, so it is a type definition, and one could expect that type definitions are "firmly known". But I wouldn't apply this to value types, since the goal of this mechanism is to track things that have been defined, rather than inferred, and this must be reflected in value types. The case that succeeds contains an explicit type annotation, and this is the intended behavior (i.e., the user has given an explicit parameter order somewhere, other than through applications). There may also be principality questions, but this is independent of the ambiguity problem the commutativity flag solves.

> As I understand it, you're refuting this statement: ...

I'm not sure what is the intent of this statement, but I just meant that the original idea behind this assertion is intuitively reasonable (and it is strange that a type definition would contain variables or unknowns), but enforcing it would require first to make it true.

Looking again at the concerned code in `build_subtype`, it would probably be fine, and more correct, to keep the original commutation flag in the generated type. It may result in hard to explain errors (since the commutativity flag is never printed), but they would be already hard to explain writing the corresponding type annotation (without subtyping).
@Octachron 's example actually makes me think this could be better, as it shows than one can leak inferred information into  the type definition environment without any explicit definition.

**Comment by Octachron** [comment]
Notwithstanding the assertion issue, there seems to be something strange going on with coercion and labels commutation:
```ocaml
module type T = sig type t  end
let type_of (type x) (_: x -> int) = (module struct type t = x end: T with type t = x)

module With_variant = struct
  type t = [`A]
  let a : t = `A
  let b : t = `A
end

module With_int = struct
  type t = int
  let a : t = 0
  let b : t = 0
end

open With_int
let f g = 1 + g ~x:a ~y:b
module Type_with_int = (val type_of f)
let commute_or_not (g:x:t -> y:t -> int) x y =
  (fun _ -> g:> 'a -> Type_with_int.t) () ~y ~x

open With_variant
(* the code below is the same as before *)
let f g = 1 + g ~x:a ~y:b
module Type_with_variant = (val type_of f)
let commute_or_not (g:x:t -> y:t -> int) x y =
  (fun _ -> g:> 'a -> Type_with_variant.t) () ~y ~x
```
It is surprising to me that half of the `commute_or_not` functions in the above example type checks.

**Comment by garrigue** [comment]
If you look at the code in `build_subtype`, you will see the reason: if there is no subtyping on both sides of the arrow, the original type is kept.
This is what happens with `With_int`, and since this type is then unified with the type of `g`, which is given explicitly, the resulting type commutes.
In the `Type_with_variant` case, a subtype is built, and unified with the explicit type, but the returned type is the non-commutative one.
So, as I wrote above, this kind of error can be confusing, but you should admit that the way you build `Type_with_variant.t` is highly non standard.

Another way to see at what happens, when using `Type_with_int.t` we are able to prove that choice of order is non-ambiguous, but when using `Type_with_variant.t`, we aren't. So the behavior is conservative.

There is still a debate on whether we should have a type error or just a warning. Ambiguity is bad, but it is not unsound.

**Comment by garrigue** [comment]
Actually, this is an extra argument for sharing the commutativity flag between the original type and the subtype: in this case this would allow the `With_variant` case to type, which is correct since subtyping cannot change the order of arguments.

**Comment by Octachron** [comment]
Those example are certainly utterly artificial, and I wrote them by looking at the code of `build_subtype`. Arguably, the error might appear in slightly less absurdly contrived situations on the object side:

```ocaml
type 'x m = 'm constraint 'x = <m:'m; ..>
module I = struct
  open Int
  class  ['a] c = object(self)
    method m f = ((f ~x ~y): unit); f
  end
  class virtual user = object(self)
    method commute_or_not (f: x:t -> y:t -> unit) =
      (fun _ -> f :> _ c m) ~y ~x
  end
end
module V = struct
  open Variant
  class  ['a] c = object
    method m f = (f ~x ~y :unit); f
  end
  class virtual user = object(self)
    method commute_or_not (f: x:t -> y:t -> unit) =
      (fun _ -> f :> _ c m) ~y ~x
  end
end
```
But yes, those examples are mostly here to illustrate an edge case.
The behavior with a shared commutativity flag seems simpler to explain.
I will open a PR with this change and the corresponding examples in the testsuite.

**Comment by Octachron** [comment]
The bug is fixed in 4.11 and trunk. I am closing this specific issue. And I will reopen a specific issue about the propagation of Cok/Cunknowns arrows in coercions.
